src/containers/atomic_wf_queue.hpp

```diff
@@ -0,0 +1,359 @@
+   * the AtomicWFQueue's internal variables such as head, tail, current size,
+   * and maximum size to their default or zero-initialized values.
+   */
+  __host__ void deallocate_queue();
+
+  /**
+   * @brief Retrieves the logical lane ID of the calling thread.
+   *
+   * This function returns the active logical lane ID of the current thread
+   * within the wavefront. The logical lane ID uniquely identifies
+   * the thread's position among active threads in the wavefront.
+   *
+   * @return The logical lane ID of the active thread within the wavefront.
+   */
+  __device__ unsigned int active_logical_lane_id();
+
+  /**
+   * @brief Broadcasts a value to other threads in the wavefront.
+   *
+   * This function broadcasts the specified value to all active threads
+   * in the wavefront. If `lowest_active` is true, the value is broadcasted
+   * from the thread with the lowest active lane ID.
+   *
+   * @param lowest_active If true, broadcasting starts from the lowest 
+   *                      active thread in the wavefront.
+   * @param val The value to be broadcasted.
+   *
+   * @return The broadcasted value received by each thread in the wavefront.
+   */
+  __device__ TYPE broadcast_lds(bool lowest_active, TYPE val);
+
+  /**
+   * @brief Retrieves the maximum capacity of the AtomicWFQueue.
+   *
+   * This function returns the total size of the AtomicWFQueue, representing
+   * the maximum number of elements it can hold.
+   *
+   * @return The maximum capacity of the AtomicWFQueue.
+   */
+  __host__ __device__ int get_queue_size() {
+    return size_;
+  }
+
+  /**
+   * @brief Retrieves the current number of elements in the AtomicWFQueue.
+   *
+   * This function returns the current size of the AtomicWFQueue, representing
+   * the total number of elements currently stored.
+   *
+   * @return The current number of elements in the AtomicWFQueue.
+   */
+  __host__ __device__ int get_curr_size() {
+    return curr_size_;
+  }  
+
+  /**
+   * @brief Retrieves the tail index of the AtomicWFQueue.
+   *
+   * This function returns the current index of the tail in the
+   * AtomicWFQueue, which represents the position where the next
+   * element will be enqueued.
+   *
+   * @return The index of the tail in the AtomicWFQueue.
+   */
+  __host__ __device__ int get_tail() {
+    return tail_;
+  }
+
+  /**
+   * @brief Retrieves the head index of the AtomicWFQueue.
+   *
+   * This function returns the current index of the head in the
+   * AtomicWFQueue, which represents the position of the next element
+   * to be dequeued.
+   *
+   * @return The index of the head in the AtomicWFQueue.
+   */
+  __host__ __device__ int get_head() {
+    return head_;
+  }
+
+ private:
+
+  __device__ int atomic_load(const int* address) {
+    return __hip_atomic_load(address, __ATOMIC_SEQ_CST,
+                             __HIP_MEMORY_SCOPE_AGENT);
+  }
+
+  __device__ void atomic_store(int* address, int val) {
```

Comment:
NIT: since you used `const int* address` above for the atomic load, shouldn't it also be `const int val` here?